routes/security.py

```diff
@@ -0,0 +1,106 @@
+from datetime import timedelta
+import logging
+from fastapi import APIRouter, Depends, HTTPException, status
+from fastapi.responses import RedirectResponse
+from dependency_injector.wiring import Provide, inject
+import jwt
+from jwt import PyJWKClient
+from models.security import AuthToken, OAuthBody, ScopeEnum
+
+from utils.container import Container
+from utils.encryption import create_jwtoken
+
+
+router = APIRouter()
+
+
+@router.post("/authorize")
+@inject
+def authorize_redirect(config=Depends(Provide[Container.config])):
+    return RedirectResponse(config["auth_url"])
+
+
+@router.post("/token", response_model=AuthToken)
+@inject
+async def token_request(
+    data: OAuthBody,
+    config=Depends(Provide[Container.config]),
+):
+    if data.grant_type == "authorization_code":
+        if data.code is None:
+            logging.error(f"Missing authorization code")
+            raise HTTPException(
+                status_code=status.HTTP_400_BAD_REQUEST,
+                detail="Missing authorization code",
+            )
+
+        try:
+            jwks_client = PyJWKClient(config["jwk_url"])
+            signing_key = jwks_client.get_signing_keys()[0]
+            jwt.decode(
+                data.code,
+                signing_key.key,
+                algorithms=["RS256"],
+                options={"require": ["exp"]},
+            )
+        except Exception:
+            logging.error("Invalid authorization code")
+            raise HTTPException(
+                status_code=status.HTTP_401_UNAUTHORIZED,
```

Comment:
When the client credentials are invalid, consider returning a 403 status code (Forbidden) instead of 401 (Unauthorized).
Please see : [https://developer.mozilla.org/fr/docs/Web/HTTP/Status/403](https://developer.mozilla.org/fr/docs/Web/HTTP/Status/403)

Reply:
Invalid credentials should return 401 as explained by Auth0 https://auth0.com/blog/forbidden-unauthorized-http-status-codes/#Recap
However it should be 403 when the scopes are not good in routes/middlewares/security.py:check_jwt which I made a 401 too. Changing it right now